### ISSUE TITLE
.. in package-env files are incompatible with persistent workers

### ISSUE THREAD

**Issue by facundominguez** [tweag/rules_haskell#1748]
I discovered today that if I build a module with the persistent worker and use a package environment file
```
-package-env A/B/package-env
```
and package-env contains relative paths like
```
package-db ../C
```
GHC will store the location of the package as `A/B/../C` which is considered by GHC the same location as `A/C`, but only for as long as the `B` directory exists.

When compiling modules with the persistent worker, the directory `B` might be removed before compiling another module, invalidating the paths that are stored in the internal state of the GHC API.

I tested that passing `-package-db A/C` in the command line fixes the issue, which leaves me with the question of how to fix `rules_haskell` to get rid of the `..` in `package-db` arguments. I'm conjecturing that `package-env` files are a solution to keep command line arguments shorter.

**Comment by facundominguez** [comment]
My best solution till now is to have the ghc wrapper script move the env file a few directories upwards and translate the `package-db` paths to eliminate the leading `..`s.

**Comment by facundominguez** [comment]
I found another solution that is to limit each persistent worker to only compile one package. Presumably, the sandboxes to compile each module of a package are mostly invariant, which would reduce the risk of the GHC caches being invalidated by changes in the sandbox. We still can allow bazel to use multiple persistent workers to compile each package and it all would be limited by the limit of concurrent jobs.

**Comment by aherrmann** [comment]
> I'm conjecturing that `package-env` files are a solution to keep command line arguments shorter.

Correct, that's why they were introduced. They are generated [here](https://github.com/tweag/rules_haskell/blob/e1017b429526e6549ef327e80ce7b3f3af3470a3/haskell/private/packages.bzl#L39-L52).

> GHC will store the location of the package as `A/B/../C` which is considered by GHC the same location as `A/C`, but only for as long as the `B` directory exists.

Could a later loaded, different package-env file overwrite this mapping? If a later build action still requires `A/C`, but doesn't have the directory `A/B/`, then it should have some other valid package-env file or direct path to `A/C`. Could that new path replace the old entry that references `A/B/../C`?

> I found another solution that is to limit each persistent worker to only compile one package. Presumably, the sandboxes to compile each module of a package are mostly invariant, which would reduce the risk of the GHC caches being invalidated by changes in the sandbox. We still can allow bazel to use multiple persistent workers to compile each package and it all would be limited by the limit of concurrent jobs.

I'm not sure I understand this. How would Bazel know that a particular persistent worker instance is only intended for a particular package?
